pkg/v2/tkr/webhook/cluster/vsphere-template-resolver/templateresolver/resolver_impl.go

```diff
@@ -0,0 +1,205 @@
+	"github.com/vmware-tanzu/tanzu-framework/pkg/v1/tkg/vc"
+)
+
+type Resolver struct {
+	VCClient vc.Client
+	Log      logr.Logger
+}
+
+func NewResolver(log logr.Logger) *Resolver {
+	return &Resolver{Log: log}
+}
+
+func (r *Resolver) InjectVCClient(vcClient vc.Client) {
+	r.VCClient = vcClient
+}
+
+// getVSphereEndpoint gets vsphere client based on credentials set in config variables
+func (r *Resolver) GetVSphereEndpoint(svrContext VSphereContext) (vc.Client, error) {
+	host := strings.TrimSpace(svrContext.Server)
+	if !strings.HasPrefix(host, "http") {
+		host = "https://" + host
+	}
+	vcURL, err := url.Parse(host)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to parse vc host")
+	}
+	vcURL.Path = "/sdk"
+	vcClient, err := vc.NewClient(vcURL, svrContext.TLSThumbprint, svrContext.InsecureSkipVerify)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create vc client")
+	}
+	_, err = vcClient.Login(context.TODO(), svrContext.Username, svrContext.Password)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to login to vSphere")
+	}
+	return vcClient, nil
+}
+
+// Resolve queries VC for template resolution of the OVAs from the input query.
+// It is guaranteed that every `TemplateQuery` in the input `query` will contain a corresponding entry in the result.
+// If a TemplateQuery in the input was empty (has a zero-length OVA Version), the corresponding result for that query will be an empty `OVATemplateResult`.
+func (r *Resolver) Resolve(svrContext VSphereContext, query Query) Result {
+	var err error
+	// 1. Find the union of the non-empty OVAVersion to be queried in vSphere(both control-plane and MDs)
+	ovas := []*types.VSphereVirtualMachine{}
+	ovas = fillVSphereVitrualMachine(ovas, query.ControlPlane)
+	ovas = fillVSphereVitrualMachine(ovas, query.MachineDeployments)
+
+	// Query VC to get templates for OVAs.
+	if len(ovas) > 0 {
+		// Only query if there are non-empty queries to resolve.
+		ovas, err = r.getVirtualMachineTemplateForOVAs(ovas, svrContext.DataCenter)
+		if err != nil {
+			return Result{UsefulErrorMessage: err.Error()}
+		}
+	}
+
+	// Populate results regardless of whether there are empty queries.
+	// Empty queries will have corresponding empty result elements.
+	result := Result{}
+	result.ControlPlane, err = r.calculateOVATemplateResult(query.ControlPlane, ovas)
+	if err != nil {
+		return Result{UsefulErrorMessage: err.Error()}
+	}
+	result.MachineDeployments, err = r.calculateOVATemplateResult(query.MachineDeployments, ovas)
+	if err != nil {
+		return Result{UsefulErrorMessage: err.Error()}
+	}
+
+	return result
+}
+
+// fillVSphereVitrualMachine returns an appended slice of `VSphereVirtualMachine` with OVA and OSInfo populated.
+// A `VSphereVirtualMachine` will not be added for entries with empty queries (If the OVA Version is empty).
+func fillVSphereVitrualMachine(ovas []*types.VSphereVirtualMachine, queries []*TemplateQuery) []*types.VSphereVirtualMachine {
```

Comment:
Thanks for catching the typo! Addressed in 660a42b7f0213d0dd9c235d395476c0591e13239.

Reply:
The second comment, I will take as a separate one as this seems to be a bit of a bigger change.

Reply:
ccing @prkalle for this - 

> Also, ovas and the return type could be map[TemplateQuery]*types.VSphereVirtualMachine: this would greatly simplify the code where it is used.